lib/lte_link_control/lte_lc.c

```diff
@@ -632,20 +632,45 @@ static int connect_lte(bool blocking)
 	int err;
 	int tries = (IS_ENABLED(CONFIG_LTE_NETWORK_USE_FALLBACK) ? 2 : 1);
 	enum lte_lc_func_mode current_func_mode;
+	enum lte_lc_nw_reg_status reg_status;
+	static atomic_t in_progress;
 
 	if (!is_initialized) {
 		LOG_ERR("The LTE link controller is not initialized");
 		return -EPERM;
 	}
 
+	err = lte_lc_nw_reg_status_get(&reg_status);
+	if (err) {
+		LOG_ERR("Failed to get current registration status");
+		return -EFAULT;
+	}
+
+	/* Do not attempt to register with an LTE network if the device already is registered.
+	 * This check is needed for blocking _connect() calls to avoid hanging for
+	 * CONFIG_LTE_NETWORK_TIMEOUT seconds waiting for a semaphore that will not be given.
+	 */
+	if ((reg_status == LTE_LC_NW_REG_REGISTERED_HOME) ||
+	    (reg_status == LTE_LC_NW_REG_REGISTERED_ROAMING)) {
```

Comment:
In `at_handler_cereg()`, also the cell ID validity is checked. Even if the device is registered to home or roaming network, if the cell ID is not valid, the registration status remains "unknown" and the semaphore is not given. This cell ID check is not done in `lte_lc_nw_reg_status_get()`, so here device would be considered registered also when the cell ID is invalid.

Should the cell ID validity be checked also here? And what about `lte_lc_nw_reg_status_get()`, is it ok that it returns `LTE_LC_NW_REG_REGISTERED_HOME` or `LTE_LC_NW_REG_REGISTERED_ROAMING` also when the cell ID is not valid?

Reply:
Discussed offline. Check for valid cell ID has been added to `lte_lc_nw_reg_status_get()`. This is needed because older modem fws may report an invalid combination of home/roaming status with no valid cell ID.